clients/src/main/java/org/apache/kafka/clients/consumer/internals/RequestFuture.java

```
@@ -37,86 +37,94 @@
+        return (T) result.get();
     }
 
     /**
      * Check if the request succeeded;
      * @return true if the request completed and was successful
      */
     public boolean succeeded() {
-        return isDone && exception == null;
+        return isDone() && !failed();
     }
 
     /**
      * Check if the request failed.
      * @return true if the request completed with a failure
      */
     public boolean failed() {
-        return isDone && exception != null;
+        return result.get() instanceof RuntimeException;
     }
 
     /**
      * Check if the request is retriable (convenience method for checking if
      * the exception is an instance of {@link RetriableException}.
      * @return true if it is retriable, false otherwise
+     * @throws IllegalStateException if the future is not complete or completed successfully
      */
     public boolean isRetriable() {
-        return exception instanceof RetriableException;
+        return exception() instanceof RetriableException;
     }
 
     /**
      * Get the exception from a failed result (only available if the request failed)
-     * @return The exception if it exists or null
+     * @return the exception set in {@link #raise(RuntimeException)}
+     * @throws IllegalStateException if the future is not complete or completed successfully
      */
     public RuntimeException exception() {
-        return exception;
+        if (!failed())
+            throw new IllegalStateException("Attempt to retrieve exception from future which hasn't failed");
+        return (RuntimeException) result.get();
     }
 
     /**
      * Complete the request successfully. After this call, {@link #succeeded()} will return true
      * and the value can be obtained through {@link #value()}.
      * @param value corresponding value (or null if there is none)
+     * @throws IllegalStateException if the future has already been completed
      */
     public void complete(T value) {
```

Maybe we should throw an exception if `value` is an instance of `RuntimeException` and specify that in the docs.

---

Sure. I thought about adding this, but figured that would be an unlikely usage. Still probably better to be safe.